Major Refactor: Monorepo

GregRos opened the pull request:
This is a major refactor that turns the repository into a monorepo together with the [char-info](https://github.com/GregRos/char-info) package.

Partial list of changes:
- Created monorepo structure
- Set up yarn workspaces
- Merged with char-info
- Modified GitHub workflows
- Set up hierarchical tsconfig.json structure
- Set up testing for char-info (it uses ava)
- Updated engines based on @matlin's #118
- Fixed up linting configuration
- Added README at monorepo root

There are a few kinks right now, though. 

## Examples structure
The structure of the examples folder mirrors that of the package, with `src` and `spec` folders. I'm not sure this is the best structure for an examples repo. Ideally you should get to the files as quickly as possible. 

✨ **I do usually like to separate test and non-test files, but I think some improvement can be made here.**

## Parjs test tools duplicate code
I copied the file that sets up the Jest matchers from the `parjs/spec` folder to the `parjs/examples/spec` folder due to compilation issues.

✨ **I think a good solution would be to have parser testing tools as a separate package that can be used by others to test their parsers.**

## Char-info tests
A weird thing is that the `parjs` spec configuration doesn't emit, but the `char-info` spec  configuration does emit into `dist-spec`. That's because `char-info` is tested with an old version of `ava` that didn't want to execute the TypeScript files directly. 

In the near future I'm going to rewrite the package and probably the tests, so I just needed to get it to work.

## typedoc
Typedoc is the tool used by Parjs to generate documentation. I've decided to stop using it for the next version and have a full-on site. I even bought [parjs.org](http://parjs.org). So I didn't put in the extra effort to get it to work in this setup.

The documentation will only be for v2, which is planned to have a number of API changes. I haven't touched the existing GitHub pages site that contains the documentation, so it should continue pointing to what was generated for an older version.

Any changes made to the API till then, before v2, won't be there unless the typedoc integration is fixed again.

# Engines Update
This also includes @matlin's #118 update to the `engines` field in the `package.json`

mikavilpas commented on package.json:
```diff
@@ -1,169 +1,50 @@
-    "char-info": "0.3.2"
+    "char-info": "workspace:^",
+    "espree": "^10.0.1",
+    "parjs": "workspace:^"
```
oh nice, I didn't know you could do this 👍🏻 

I work in a project or two at work where this is done, sometimes poorly, but this looks quite clean.

GregRos replied:
Thank you! I've tried to do it properly :)

mikavilpas commented on package.json:
```diff
@@ -172,10 +53,10 @@
+  "packageManager": "yarn@4.2.2",
   "engines": {
-    "node": "^18.13.0 || ^20.9.0"
+    "node": ">=16.0.0"
```
suggestion: I don't have a preference for the node version, but right now `.nvmrc` specifies node 20. I'm thinking we could switch that to 16 as well.

GregRos replied:
I'd rather not, for a few reasons.

- Node 16 is less stable and performs worse than 20 overall, so it would make development more annoying.
- The current version of Yarn requires 18+, and so we'd need to downgrade it a lot. Maybe all the way down to 1.x. Then we'd need to regenerate the lockfile, change configuration, some of the CLI has also changed, etc.
- Profiling was unstable on Node 16, though not as bad as on Node 14. But it works great on Node 20.
- ESM won't work properly.
- A lot of popular packages have already dropped support for 16.

mikavilpas replied:
These sound like good reasons. But is there something we could do (or not do) to ensure all these node versions are supported?

I don't think this is an issue - I'm wondering what would be a good way of working.

mikavilpas commented on packages/parjs/spec/unit/combinators/expects.spec.ts:
```diff
@@ -1,5 +1,5 @@
+import { reason, then } from "@lib/combinators";
 
 describe("expects combinator", () => {
     const base = nope("deez nuts");
```
nice

mikavilpas commented:
Nice, these changes look like they're going in a great direction.

Some additional suggestions:
- will this be a breaking change? Let's remember to bump the major version before releasing.
- there's some additional documentation under `documentation/` that probably needs updating. For example, I put a simple contribution guide that probably overlaps with the upcoming guide - maybe the old one could simply be deleted.

GregRos commented on README.md:
````diff
@@ -1,429 +1,177 @@
+└── .git*
+```
+
+###
````
Nice catch! Haha

GregRos commented on README.md:
```diff
@@ -1,429 +1,177 @@
+
+In this case, it's a dev dependency. It seems to be needed in each package.
+
+### TypeScript Projects
+
+| Symbol | Meaning                                                                                   |
+| ------ | ----------------------------------------------------------------------------------------- |
+| ⋆      | `tsconfig.json` that compiles and emits JavaScript.                                       |
+| ⋄      | `tsconfig.json` that doesn't emit for tests that should be run with `ts-node` or similar. |
+| ∘      | `tsconfig.json` that only has references to other projects and no files.                  |
+
+TypeScript project at ∘`/tsconfig.json`, referencing:
+
+-   `/packages/parjs/tsconfig.json`, referencing
+    -   ⋆ `/packages/parjs/src/tsconfig.json`
+    -   ⋄ `/packages/parjs/spec/tsconfig.json`
+    -   ∘ `/packages/parjs/examples/tsconfig.json`, referencing
+        -   ⋆ `/packages/parjs/examples/src/tsconfig.json`
+        -   ⋄ `/packages/parjs/examples/spec/tsconfig.json`
+-   ∘ `/packages/char-info/tsconfig.json`, referencing
+    -   ⋆ `/packages/char-info/src/tsconfig.json`
+    -   ⋆ `/packages/char-info/spec/tsconfig.json
+
+⋆ and ⋄ tsconfigs extend `/tsconfig.base.json`, which allows it to configure compilation for the whole project. The only other properties in the different `tsconfig`s are project-specific and set `noEmit`, `composite`, `paths`, and so on. There are no meaningful overrides.d
```
That's a great idea!

GregRos commented on packages/char-info/README.md:
```diff
@@ -0,0 +1,61 @@
+[![Codecov](https://codecov.io/gh/GregRos/char-info/branch/master/graph/badge.svg)](https://codecov.io/gh/GregRos/char-info/branch/master)
+[![NPM](https://badge.fury.io/js/char-info.svg)](https://www.npmjs.com/package/char-info)
+
+[API Documentation](https://char-info.netlify.com/)
```
Hmm it's not exactly to embed the library inside Parjs. It's to have them in the same repo. They are separate packages that are managed in the same repository.

But yeah, this is all messed up. I tried not touch it too much because it needs a complete rewrite more or less, but these links can be updated.

GregRos commented on packages/char-info/README.md:
```diff
@@ -0,0 +1,61 @@
+# char-info
+
+[![Travic CI](https://travis-ci.org/GregRos/char-info.svg?branch=master)](https://travis-ci.org/GregRos/char-info)
```
I updated the package to use GitHub actions but I forgot to update the readme :smile:

GregRos replied:
I think I need to run the push workflow for the badge to be produced.

GregRos commented on README.md:
```diff
@@ -1,429 +1,177 @@
 
--   [API Documentation](https://gregros.github.io/parjs/)
--   [Using Parjs - Developer guide](/documentation/using-parjs.md)
+The rest of this is going to be developer documentation.
```
This is the readme file for the repository, rather than the package, so I thought it should have developer documentation. 

The file `./packages/parjs/README.md`, which is the readme for the `parjs` package, is meant to be user-facing. It's what will be seen by users viewing the package on `npmjs.com`, when it's published.

However, the repo is called `parjs`, after the package, and this readme is the first thing some users will see. So I think you have a point. Let's see what other monorepos put in their root readme.

- [rxjs](https://github.com/ReactiveX/rxjs) only has developer documentation.
- [react](https://github.com/facebook/react) is mostly user-facing.
- [parcel](https://github.com/parcel-bundler/parcel) has a symbolic link to the readme of the main package. So user-facing.
- [nx](https://github.com/nrwl/nx) has links to user-facing documentation.
- [yarn](https://github.com/yarnpkg/berry) has links to user-facing and some developer.
- [this one](https://github.com/ethereumjs/ethereumjs-monorepo) only has developer documentation.

So there isn't a single answer. I'll think about it some more.

GregRos replied:
I ended up just copy-pasting the package README and putting all the developer stuff in the end.

Logically, I thought that the README for the repository root should be about a repository, rather than its main component. But that's outweighed by more practical considerations.

This readme is a non-neglible web presence. So even if there is a website with documentation, which there is going to be, it's still a waste not to use that real estate to show off how awesome the library is.

I'm not sure if this is the best way to use that real estate, but because v2 will have major changes and way more documentation, it seems a bit of a waste to work too hard on this readme right now.

GregRos commented:
> Nice, these changes look like they're going in a great direction.
> 
> Some additional suggestions:
> 
>     * will this be a breaking change? Let's remember to bump the major version before releasing.
> 
>     * there's some additional documentation under `documentation/` that probably needs updating. For example, I put a simple contribution guide that probably overlaps with the upcoming guide - maybe the old one could simply be deleted.

I tried quite hard to make sure there are no changes at all, not even minor ones. 

So for the documentation, not sure at this point.

GregRos commented on packages/char-info/src/names/categories.ts:
```diff
@@ -0,0 +1,50 @@
+
+    export const PunctuationConnector = "Pc";
+    export const PunctuationDash = "Pd";
+    export const PuncutationOpen = "Ps";
```
Yes, you're right, but sadly I can't change it because it's part of the public API of the package. 

The whole package will be changed radically in v2 though, including this part.

GregRos commented:
I think I've finished addressing all your comments @mikavilpas. I'm gonna try to merge this. And then I'll do lots of manual tests to make sure everything works.
